### PR TITLE
CAT-2346 Right-to-Left support for layouts and drawables and categories, minsdk = 17

### PR DESCRIPTION
**Pull request by ndrnour**
The UI of an RTL-language is generally mirrored and right aligned.Naturally, the common reading order for the speakers of RTL-languages is from RTL. That is because those languages are written in a form known as RTL and strings flows in that direction

### CODE REVIEW

**Comment by 84n4n4** [comment]
please run espresso tests

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+				.check(matches(isCompletelyDisplayed()));
+		onView(withIndex(withId(R.id.scratch_projects_list_item_details_text), 1))
+				.check(matches(isCompletelyDisplayed()));
+		onView(withIndex(withId(R.id.scratch_projects_list_item_title), 1))
+				.check(isLeftOf(withIndex(withId(R.id.scratch_projects_list_item_image), 1)));
+		onView(withIndex(withId(R.id.scratch_projects_list_item_details_text), 1))
+				.check(isLeftOf(withIndex(withId(R.id.scratch_projects_list_item_image), 1)));
+	}
+
+	private boolean checkViewDirection(int viewId) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java]: please do this via implementing a custom matcher.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+		return ViewCompat.LAYOUT_DIRECTION_RTL == ViewCompat.getLayoutDirection(baseActivityTestRule.getActivity()
+				.findViewById(viewId));
+	}
+
+	private boolean checkTextDirection(String string) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java]: ive seen this function with another name already in your other prs, same applies here.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+		assertTrue(checkViewDirection(R.id.main_menu_button_upload));
+
+		assertTrue(checkTextDirection(getResourcesString(R.string.main_menu_continue)));
+		assertTrue(checkTextDirection(getResourcesString(R.string.main_menu_new)));
+		assertTrue(checkTextDirection(getResourcesString(R.string.main_menu_programs)));
+		assertTrue(checkTextDirection(getResourcesString(R.string.main_menu_help)));
+		assertTrue(checkTextDirection(getResourcesString(R.string.main_menu_web)));
+		assertTrue(checkTextDirection(getResourcesString(R.string.main_menu_upload)));
+
+		onView(withText(R.string.main_menu_programs))
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java]: do not move through activites / perform actions you actually are not testing.
if you want to check the program menu activity, or other activites (like look list activity further down). write separate test classes for those and open them using the activity test rule.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+		onView(withIndex(withId(R.id.spritelist_item_background), 0))
+				.perform(click());
+		onView(withText(R.string.scripts))
+				.check(matches(isCompletelyDisplayed()));
+		onView(withText(R.string.backgrounds))
+				.check(matches(isCompletelyDisplayed()));
+		onView(withText(R.string.sounds))
+				.check(matches(isCompletelyDisplayed()));
+
+		assertTrue(checkTextDirection(getResourcesString(R.string.scripts)));
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java]: is there any specific reasons that those asserts are here in a UI test?
and does it make sense in the first place to check if every string is translated?

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/matchers/ViewMatchers.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeMatcher;
+
+public final class ViewMatchers {
+	private ViewMatchers() {
+		throw new AssertionError();
+	}
+
+	public static Matcher<View> withIndex(final Matcher<View> matcher, final int index) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/matchers/ViewMatchers.java]: can you please explain what this is supposed to do / how this works?

**Comment by ndrnour** [comment]
what's the name of this Activity?

dump_6846307136205197319

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java (hunk 1 above)
Comment (ndrnour) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java]: done

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java (hunk 2 above)
Comment (ndrnour) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java]: done

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/matchers/ViewMatchers.java (hunk 5 above)
Comment (ndrnour) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/matchers/ViewMatchers.java]: done

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java (hunk 3 above)
Comment (ndrnour) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java]: ok, I just wrote 2 test cases for MainMenuActivity and MyProjectsActivity. after the 5 Rtl PRs get merged, I think should I or someone else writes an RTL test cases for the whole App

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java (hunk 4 above)
Comment (ndrnour) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/activity/rtl/RtlLayoutsTest.java]: > is there any specific reasons that those asserts are here in a UI test?

those asserts are replaced with a custom matcher

> and does it make sense in the first place to check if every string is translated?

could you be specific with your Question?

**Comment by cheidenreich** [comment]
same question as on #2325

**Comment by 84n4n4** [comment]
test this please

**Comment by 84n4n4** [comment]
test this please

**Comment by 84n4n4** [comment]
summary:
- upgrade to api17
- changing paddingLeft to paddingStart, paddingRight to paddingEnd, etc. to make the same layout xmls work for both directions ltr and rtl.
- added icons and other drawables mirrored (because arrows supposed to be pointing other direction, question marks the other way round, etc...)
- added labelFor to layouts, that is required by lint checks for api >= 17. this is for android accessibility stuff, where it is required to state if a textView is the label for a edit text.

@cheidenreich @thmq can you please also have a look?

**Comment by 84n4n4** [comment]
please run espresso tests

**Pull request merged**